netx/conn.go

```diff
@@ -0,0 +1,138 @@
+package netx
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/sirupsen/logrus"
+)
+
+// InstanceProvider is the channel where updated list of desired service are published
+type InstanceProvider <-chan []Address
+
+// Sender sends payload to given Address
+type Sender func(addr Address, payload []byte) (int, error)
+
+// Address of a service in IP:PORT format
+type Address string
+
+// RoundRobin returns writer with round robin functionality. Every write could be sent to different backend.
+func RoundRobin(instanceProvider InstanceProvider, sender Sender) io.Writer {
+	return &roundRobinWriter{provider: instanceProvider, sender: sender, instances: nil}
+}
+
+type roundRobinWriter struct {
+	provider  InstanceProvider
+	sender    Sender
+	instances chan Address
+}
+
+func (r *roundRobinWriter) Write(byte []byte) (int, error) {
+
+	if r.instances == nil {
+		r.updateInstances(<-r.provider)
+	}
+
+	select {
+	case newInstances := <-r.provider:
+		r.updateInstances(newInstances)
+		return r.write(byte)
+	default:
+		return r.write(byte)
+	}
+}
+
+func (r *roundRobinWriter) updateInstances(newInstances []Address) {
+	r.instances = make(chan Address, len(newInstances))
+	for _, instance := range newInstances {
+		r.instances <- instance
+	}
+}
+
+func (r *roundRobinWriter) write(payload []byte) (int, error) {
+	// Read next instance from queue
+	instance := <-r.instances
+	// Enqueue instance for round robin behaviour
+	r.instances <- instance
+
+	return r.sender(instance, payload)
+}
+
+// NetworkSend is a Sender that can write payload to the network Address
+func NetworkSend(addr Address, payload []byte) (int, error) {
+	conn, err := net.Dial("udp", string(addr))
```

Comment:
For UDP case it will be more optimal to use a `UDPConn.WriteTo` method, without creating (and later closing) virtual connection.